src/visGeometry/index.ts

```diff
@@ -1542,51 +1553,49 @@ class VisGeometry {
             lastz = 0;
 
         this.legacyRenderer.beginUpdate(this.scene);
-
         this.fibers.beginUpdate();
         this.geometryStore.forEachMesh((agentGeo) => {
             agentGeo.geometry.instances.beginUpdate();
         });
-        // these lists must be emptied on every scene update.
+
+        // Clear draw lists
         this.agentsWithPdbsToDraw = [];
         this.agentPdbsToDraw = [];
 
-        // First, mark ALL inactive and invisible.
-        // Note this implies a memory leak of sorts:
-        // the number of agent instances can only grow during one trajectory run.
-        // We just hide the unused ones.
-        // Worst case is if each frame uses completely different (incrementing) instance ids.
-        for (let i = 0; i < MAX_MESHES && i < this.visAgents.length; i += 1) {
-            const visAgent = this.visAgents[i];
-            visAgent.hideAndDeactivate();
+        // Mark all agents as inactive and invisible
+        for (let i = 0; i < MAX_MESHES && i < this.visAgents.length; i++) {
+            this.visAgents[i].hideAndDeactivate();
         }
 
-        agents.forEach((agentData) => {
+        let offset = AGENT_HEADER_SIZE;
+        const newVisAgentInstances = new Map<number, VisAgent>();
+        for (let i = 0; i < agentCount; i++) {
+            const agentData = getAgentDataFromBuffer(view, offset);
             const visType = agentData.visType;
             const instanceId = agentData.instanceId;
             const typeId = agentData.type;
+
             lastx = agentData.x;
             lasty = agentData.y;
             lastz = agentData.z;
 
-            // look up last agent with this instanceId.
             let visAgent = this.visAgentInstances.get(instanceId);
 
             const path = this.findPathForAgent(instanceId);
-            if (path) {
-                if (visAgent) {
-                    lastx = visAgent.agentData.x;
-                    lasty = visAgent.agentData.y;
-                    lastz = visAgent.agentData.z;
-                }
+            if (path && visAgent) {
+                lastx = visAgent.agentData.x;
+                lasty = visAgent.agentData.y;
+                lastz = visAgent.agentData.z;
             }
 
             if (!visAgent) {
-                visAgent = this.createAgent();
+                if (this.availableAgentPool.length > 0) {
+                    visAgent = this.availableAgentPool.pop() as VisAgent;
```

Comment:
any concern about leftover data in this visAgent, or does the following code completely overwrite its internals?

Reply:
Before pooling we reset any property that doesn't get explicitly set in the following code.

We could be more careful and reset everything when we pool, but since the following code should overwrite those properties (like agentData) I tried to make the pooling process as lean as possible.